### DISCUSSION TITLE
PEP or other official document about (mandatory) use of "requires_python" and "classifier" (with Python version) fields?

### DISCUSSION THREAD

**Discussion by buhtz** [pypa/setuptools#3816] [Q&A] [answered]
I know two ways to specify the required Python version.

```
[project]
# ...
requires-python = ">=3.8"
# ...
classifiers = [
    "Programming Language :: Python :: 3.8"
# ...
```

This is redundant. But for example [tools like `pyroma` make them mandatory](https://github.com/regebro/pyroma/issues/96).
I couldn't find a PEP or other official source to decide if I can leave one of that fields?

Do you have any source or opinion about it?

**Comment by CAM-Gerlach** [comment]
Briefly, classifiers are particularly important to specify the _upper bound_ that the package is actually tested to work against, without being a hard cap for the solver like `requires_python` (which causes a bunch of problems in practice). See [this discussion on the official Packaging Discourse](https://discuss.python.org/t/requires-python-upper-limits/12663) for more information and context there.

However, this would likely be more on topic for the [official Python Packaging Discourse](https://discuss.python.org/c/packaging/14) instead, since this is spec/usage advice and not anything specific to the implementation in any one particular backend (like Setuptools).

**Comment by abravalheri** [comment, marked as answer]
They are not mandatory. The only metadata that is mandatory is `name` (and somehow `version`, but version can be `dynamic`). Reference: https://packaging.python.org/en/latest/specifications/declaring-project-metadata/.

Remark: I personally think `requires-python` is preferred over `classifiers` because tools are more likely to enforce `requires-python`.
